parity-clib/examples/cpp/main.cpp

```diff
@@ -14,172 +14,136 @@
 // You should have received a copy of the GNU General Public License
 // along with Parity Ethereum.  If not, see <http://www.gnu.org/licenses/>.
 
+#ifdef __clang__
+#pragma clang diagnostic error "-Wvexing-parse"
+#endif
 #include <chrono>
+#include <cstdint>
 #include <parity.h>
+#include <parity.hpp>
 #include <regex>
 #include <string>
 #include <thread>
-
-void* parity_run(std::vector<const char*>);
-int parity_subscribe_to_websocket(void*);
-int parity_rpc_queries(void*);
-
-const int SUBSCRIPTION_ID_LEN = 18;
-const size_t TIMEOUT_ONE_MIN_AS_MILLIS = 60 * 1000;
-const unsigned int CALLBACK_RPC = 1;
-const unsigned int CALLBACK_WS = 2;
+namespace {
+using namespace std::literals::string_literals;
+using parity::ethereum::parity_subscription;
+using parity::ethereum::ParityEthereum;
+
+void parity_subscribe_to_websocket(ParityEthereum &ethereum);
+void parity_rpc_queries(ParityEthereum &);
+parity::ethereum::ParityEthereum parity_run(const std::vector<std::string> &);
```

Comment:
I am not sure how strict we are on ourselves here but I want to point something out for this API:
If I were to translate this API directly to Rust it would look something similar to this:
```rust
fn parity_run(args: &Vec<String>) -> parity::ethereum::ParityEthereum { ... }
```
If this was a Rust codebase I wouldn't accept such as an API and instead propose something like:
```rust
fn parity_run<'a, A>(args: A) -> parity::ethereum::ParityEthereum
where
    A: IntoIterator<Item = &'a str> + 'a,
{ ... }
```
So iterators instead of reference to `Vec` and `str` instead of `String` since it doesn't seem that we are going to take ownership of those `String` instances anyways. Also we might think about returning a `Result` instead but I don't know if the C++ code throws.
This Rust code has somewhat of a translation into C++ as the following:
```c++
template<typename ArgsIt>
auto parity_run(ArgsIt begin, ArgsIt end) -> parity::ethereum::ParityEthereum;
```
The good thing is that we are generic over an input range (iterator). The bad thing is that since C++ is missing concepts (before C++20) we cannot annotate type interface (e.g. Rust trait bounds) so the underlying implementation decides what is done here.

Now since in C++ templates are sometimes avoided in interfaces etc. we might want to think about another non-template solutions.

If we actually want to re-use and take ownership of the arguments we should do:
```c++
auto parity_run(args: std::vector<String>) -> parity::ethereum::ParityEthereum;
```
or
```c++
auto parity_run(args: std::vector<std::string_view>) -> parity::ethereum::ParityEthereum;
```

If we cannot use a template here we should still not use `std::vector<String>&` (reference to vec) for the same reasons we avoid doing so in Rust. What we can do instead is to simply use `std::vector<std::string_view>`. The big problem in C++ is that as until recently C++ had no type to express what is expressed in Rust through slices (e.g. `&[i32]`). There were proposals for an `std::array_view` type, but the (renamed) `std::span` type just got into C++20. So we cannot use it, yet.

If done without generics a generic way to deal with such an interface in C++ is to use a concrete iterator.
```c++
auto parity_run(begin: MyArgsIter, end: MyArgsIter) -> parity::ethereum::ParityEthereum;
```
However, this is super super ugly.
Conclusion: C++20 is the fix but it isn't there yet. In theory we should use template if we can. Otherwise we should think about the best possible interface from the above mentioned. Maybe you can come up with another one.

Reply:
> If we cannot use a template here we should still not use std::vector<String>& (reference to vec) for the same reasons we avoid doing so in Rust.

For what it's worth, `const foo&` is very idiomatic in C++. Semantically I really isn't the same as `&Foo` in Rust.